src/main/java/org/candlepin/subscriptions/marketplace/MarketplaceProducer.java

```diff
@@ -46,27 +54,93 @@
+        this.rejectedCounter = meterRegistry.counter("rhsm-subscriptions.marketplace.batch.rejected");
+        this.properties = properties;
     }
 
     @Timed("rhsm-subscriptions.marketplace.usage.submission")
     public StatusResponse submitUsageRequest(UsageRequest usageRequest) {
-        // NOTE: https://issues.redhat.com/browse/ENT-3609 will address failures
-        return retryTemplate.execute(context -> tryRequest(usageRequest));
+        try {
+            StatusResponse status = retryTemplate.execute(context -> tryRequest(usageRequest));
+            Set<String> batchIds = Optional.ofNullable(status.getData()).orElse(Collections.emptyList())
+                .stream()
+                .map(BatchStatus::getBatchId)
+                .collect(Collectors.toSet());
+            if (properties.isVerifyBatches()) {
+                verifyBatchIds(batchIds);
+            }
+            return status;
+        }
+        catch (Exception e) {
+            rejectedCounter.increment();
+            throw e;
+        }
+    }
+
+    private void verifyBatchIds(Set<String> batchIds) {
+        batchIds.forEach(batchId -> {
+            try {
+                retryTemplate.execute(context -> verifyBatchId(batchId));
+            }
+            catch (Exception e) {
+                log.error("Error verifying batchId {}", batchId, e);
+                unverifiedCounter.increment();
+            }
+        });
+    }
+
+    private String verifyBatchId(String batchId) {
+        try {
+            StatusResponse response = marketplaceService.getUsageBatchStatus(batchId);
+            String status = Objects.requireNonNull(response.getStatus());
+            if ("inprogress".equals(status)) {
+                // throw an exception so that retry logic re-checks the batch
+                throw new MarketplaceUsageSubmissionException(response.getMessage(), status);
+            }
+            else if (!"accepted".equals(status)) {
+                log.error("Marketplace rejected batch {} with status {} and message {}", batchId, status,
+                    response.getMessage());
+                rejectedCounter.increment();
+            }
+            else {
+                acceptedCounter.increment();
+            }
+            return status;
+        }
+        catch (ApiException e) {
+            throw new SubscriptionsException(
+                ErrorCode.REQUEST_PROCESSING_ERROR,
+                Response.Status.fromStatusCode(e.getCode()),
+                "Exception checking usage batch in Marketplace",
+                e
+            );
+        }
     }
 
     private StatusResponse tryRequest(UsageRequest usageRequest) {
         try {
             StatusResponse status = marketplaceService.submitUsageEvents(usageRequest);
             log.debug("Marketplace response: {}", status);
+            if (!Arrays.asList(SUCCESSFUL_SUBMISSION_STATUSES).contains(status.getStatus())) {
```

Comment:
If we are just going to convert ```SUCCESSFUL_SUBMISSION_STATUSES``` to a list each time, we might as well just create the static final List above.

```java
private static final List<String> SUCCESSFUL_SUBMISSION_STATUSES = List.of("accepted", "inprogress");
```